rust/arrow/src/buffer/mutable.rs

```diff
@@ -415,6 +415,61 @@ impl MutableBuffer {
         buffer
     }
 
+    /// Creates a [`MutableBuffer`] from a boolean [`Iterator`] with a trusted (upper) length.
+    /// # use arrow::buffer::MutableBuffer;
+    /// # Example
+    /// ```
+    /// # use arrow::buffer::MutableBuffer;
+    /// let v = vec![false, true, false];
+    /// let iter = v.iter().map(|x| *x || true);
+    /// let buffer = unsafe { MutableBuffer::from_trusted_len_iter_bool(iter) };
+    /// assert_eq!(buffer.len(), 1) // 3 booleans have 1 byte
+    /// ```
+    /// # Safety
+    /// This method assumes that the iterator's size is correct and is undefined behavior
+    /// to use it on an iterator that reports an incorrect length.
+    // This implementation is required for two reasons:
+    // 1. there is no trait `TrustedLen` in stable rust and therefore
+    //    we can't specialize `extend` for `TrustedLen` like `Vec` does.
+    // 2. `from_trusted_len_iter_bool` is faster.
+    pub unsafe fn from_trusted_len_iter_bool<I: Iterator<Item = bool>>(
+        mut iterator: I,
+    ) -> Self {
+        let (_, upper) = iterator.size_hint();
+        let upper = upper.expect("from_trusted_len_iter requires an upper limit");
+
+        let mut result = {
+            let byte_capacity: usize = upper.saturating_add(7) / 8;
+            MutableBuffer::new(byte_capacity)
+        };
+
+        'a: loop {
+            let mut byte_accum: u8 = 0;
+            let mut mask: u8 = 1;
+
+            //collect (up to) 8 bits into a byte
+            while mask != 0 {
+                if let Some(value) = iterator.next() {
+                    byte_accum |= match value {
```

Comment:
I wonder if the bool iterator could be split into chunks (for example, using https://docs.rs/itertools/0.4.2/itertools/struct.Chunks.html or alternatively using https://doc.rust-lang.org/std/primitive.slice.html#method.chunks) of 8 bool values, then each chunk is mapped into a byte by converting each bool value into a byte, then shifting according to the position in the chunk, and applying in the output byte, and finally the resulting byte iterator would be used to build the buffer directly. This is the fastest implementation I can imagine because it eliminates as many conditions / checks as possible (and conditions are the enemy of fast).

Reply:
I also think there are some faster ways to speed up the inner loop, yours sounds like a great idea to try out. I was also looking at the arrow2 repository of @jorgecarleitao , but I think I have been looking to an older commit before which turned out to be slower (I expected it to be faster, but sometimes the compiler can be quite surprising in what compiles to efficient code.

I think the latest version is over here:

https://github.com/jorgecarleitao/arrow2/blob/be905f1b1f0293ef427387bc35b2e9956ec3336f/src/bitmap/mutable.rs#L209

Reply:
I agree with you all ❤️ 

I admit I have spent an immoral amount of time trying to optimize bitmaps, but I have unfortunately not yet concluded what is the best way to handle them. I think that we may not being able to express to the compiler what we want it to do (some kind of operation over a single byte). @yordan-pavlov suggestion is a great one in that direction, though.

FWIW, on my computer (a VM on azure), arrow master (not this PR) is giving

```
eq Float32              time:   [113.40 us 114.81 us 116.28 us]
eq scalar Float32       time:   [96.824 us 98.638 us 101.34 us]
```

and `arrow2` is giving

```
eq Float32              time:   [84.519 us 86.065 us 87.772 us]
eq scalar Float32       time:   [57.682 us 58.315 us 59.014 us]
```

This PR's idea on arrow2 (with corresponding changes) is giving me `-14%` on `eq Float32` and `+35%` on `eq scalar Float32`. I pushed these benches to master there.

Reply:
FYI, I wrote a [minimal repo](https://github.com/jorgecarleitao/bitmap_perf_test) to evaluate these things and added a [reddit post](https://www.reddit.com/r/rust/comments/m9qsqe/help_performant_write_of_comparison_to_a_lsb/) to try to get some help / feedback on this.

Reply:
@yordan-pavlov's idea yields `-50%` on array-to-scalar and `-10%` on array-to-array 🚀

Reply:
I put the latest example from the example repo here https://godbolt.org/z/PKKbfq

Some things I notice in this part

```
if chunk[i] == rhs {
                    *byte = set(*byte, i)
                }
```

I am not an export at reading the output from compiler explorer, but  the inner (0..8) loop I would say could maybe benefit from unrolling in the `0..8` loop?

Reply:
```rust
            chunk.iter().enumerate().for_each(|(i, &chunk_i)| {
                if chunk_i == rhs {
                    *byte = set(*byte, i)
                }
            });
```

This seems to generate a bit more efficient code for the outer loop.

Reply:
I also tried (but got similar / slightly worse results):

```
            *byte = ((chunk[0] == rhs) as u8)
                | ((chunk[1] == rhs) as u8) << 1u8
                | ((chunk[2] == rhs) as u8) << 2u8
                | ((chunk[3] == rhs) as u8) << 3u8
                | ((chunk[4] == rhs) as u8) << 4u8
                | ((chunk[5] == rhs) as u8) << 5u8
                | ((chunk[6] == rhs) as u8) << 6u8
                | ((chunk[7] == rhs) as u8) << 7u8
```

Reply:
Last experiment, also gets similar results, suggesting at least there is not much to win from unrolling :

```rust
            if unsafe { *chunk.get_unchecked(0) == rhs } {
                *byte = set(*byte, 0)
            }
            if unsafe { *chunk.get_unchecked(1) == rhs } {
                *byte = set(*byte, 1)
            }

            if unsafe { *chunk.get_unchecked(2) == rhs } {
                *byte = set(*byte, 2)
            }

            if unsafe { *chunk.get_unchecked(3) == rhs } {
                *byte = set(*byte, 3)
            }

            if unsafe { *chunk.get_unchecked(4) == rhs } {
                *byte = set(*byte, 4)
            }

            if unsafe { *chunk.get_unchecked(5) == rhs } {
                *byte = set(*byte, 5)
            }

            if unsafe { *chunk.get_unchecked(6) == rhs } {
                *byte = set(*byte, 6)
            }

            if unsafe { *chunk.get_unchecked(7) == rhs } {
                *byte = set(*byte, 7)
            }
```

Reply:
Ah, putting `-C opt-level=3` reveals the compiler is already unrolling the code :)

Reply:
Thanks a lot, @Dandandan . Note that we are getting the same performance as if we would perform this via a `Vec<bool>` (on the benches on the example repo).

A Reddit comment was in the same direction as yours, and ended up with "use SIMD". So, I think that we may be on the edge of what we can do.

Reply:
Strange still, wouldn't expect here that a bunch of `cmp` + `je` in the fastest version would be faster than variations of this, but the benchmarks show otherwise:

```
                *byte |= if chunk[i] == rhs {
                    BIT_MASK[i]
                } else {
                    0
                };
```

but the benchmarks show otherwise.

Reply:
For reference, this is part of the output from ```if chunk[i] == rhs {*byte = set(*byte, i)}``` 

![image](https://user-images.githubusercontent.com/163737/111901287-8094c580-8a37-11eb-9d77-e572f01f9212.png)



And this without jumps:
![image](https://user-images.githubusercontent.com/163737/111901260-56db9e80-8a37-11eb-8163-a19b61d0eb0c.png)

Reply:
Ah I have an explaination @jorgecarleitao 

The benchmark in your test repo only match for 10% of the values. 
When 50% of the values match (changing the benchmark to use `%2`), this version becomes faster.
So I think it is important to see what how want to benchmark / where we want to optimize for,

```
scalar_eq_bitmap        time:   [1.3688 us 1.3720 us 1.3751 us]                              
                        change: [+63.468% +64.269% +65.058%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

scalar_eq_bitmap1       time:   [1.3381 us 1.3410 us 1.3441 us]                               
                        change: [+78.511% +79.377% +80.215%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild

scalar_eq_bitmap2       time:   [945.68 ns 947.16 ns 948.87 ns]                               
                        change: [+4.5904% +5.0361% +5.4662%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) low mild
  7 (7.00%) high mild
  3 (3.00%) high severe
```

Reply:
This seems to get rid of one check in the outer loop (no big win)

```rust
buffer[..chunks]
        .iter_mut()
        .zip(chunks_iter)
        .for_each(|(byte, chunk)| {
            chunk.iter().zip(BIT_MASK.iter()).for_each(|(&c_i, &b)| {
                *byte |= if c_i == rhs {
                    b
                } else {
                    0
                };
            });
        });
```

Reply:
@Dandandan  @jorgecarleitao it looks like this implementation merges in the same loop (1) the equality check / comparison operation and (2) setting the corresponding bit in the bitmap. I wonder what difference would it make if the comparison operation and the bitmap setting operation are done separately: first the comparison operation generates an iterator of bool, then that bool iterator is translated into a bitmap (similar to how it's done in the implementation above). 

Another interesting thing to try is to fully separate the comparison operation (so that the CPU can focus on one type of operation at a time), by materializing the comparison iterator into a `Vec` using `collect()` before using the result to generate a bitmap. 

Or may be you've already tried this. Any thoughts?

Reply:
I added a version of this as a PR to your arrow2 repository https://github.com/jorgecarleitao/arrow2/pull/17

Reply:
Turns out, this latest approach can be 5-10x faster on more realistic benchmarks (see PR)!

Reply:
@yordan-pavlov 
I am not sure where the win would be in that case.
I would expect the first idea to be compiled to roughly the same code (if all compiler optimizations work out)?

For the `Vec` one - I would expect that would be slower as it introduces an extra loop / allocation and barrier for optimization?

Reply:
@Dandandan yes, I agree that it's counter-intuitive, but I find that the compiler often surprises me so it's best to try every option; I will try to extend the benchmark repo when I have a bit more free time later today

Reply:
You are right, sometimes the result can be surprising.
I tried this variation, this compiles to the same unrolled 38 instructions:
```rust
            chunk.iter().map(|&c_i| c_i == rhs).enumerate().for_each(|(i, c_i)| {
                *byte |= if c_i {
                    1 << i
                } else {
                    0
                };
```

Reply:
yes, I tried to benchmark doing the comparison separately, but it's not faster; on my machine the fastest version is:
```
(0..8).for_each(|i| {
                if chunk[i] == rhs {
                    *byte = set(*byte, i)
                }
            });
```
and that's even faster than: 
```
chunk.iter().enumerate().for_each(|(i, &c_i)| {
                *byte |= unsafe { mem::transmute::<bool, u8>(c_i == rhs) << i };
            });
```
this is with the test data configured as:
```
let vec = (0..20049).map(|x| (x * x + x) % 2).collect::<Vec<_>>();
```
I think 2000 items (the old length of the test data) is much too small for realistic benchmarking, and it would make more sense to benchmark with test date with length same as the default batch size in DataFusion (I think this was recently increased).

Reply:
@yordan-pavlov 

Also see findings in the PR here (with updated benchmark). `(x * x + x) % 2` will give 0 on even and 1 on uneven inputs I, so the pattern/branches to include will be very predictable, especially in the scalar version.

https://github.com/jorgecarleitao/arrow2/pull/17